Classes/Private/Manager/Audio/OCTAudioEngine.m

```diff
@@ -83,459 +38,162 @@ - (instancetype)init
+    self.inputQueue.sendDataBlock = ^(void *data, OCTToxAVSampleCount samples, OCTToxAVSampleRate rate, OCTToxAVChannels channelCount) {
+        OCTAudioEngine *aoi = welf;
 
-OSStatus inputRenderCallBack(void *inRefCon,
-                             AudioUnitRenderActionFlags  *ioActionFlags,
-                             const AudioTimeStamp    *inTimeStamp,
-                             UInt32 inBusNumber,
-                             UInt32 inNumberFrames,
-                             AudioBufferList *ioData)
-{
-    OCTAudioEngine *engine = (__bridge OCTAudioEngine *)(inRefCon);
+        if (aoi.enableMicrophone) {
+            [aoi.toxav sendAudioFrame:data
+                          sampleCount:samples
+                             channels:channelCount
+                           sampleRate:rate
+                             toFriend:aoi.friendNumber
+                                error:nil];
+        }
+    };
+    [self.outputQueue updateSampleRate:(Float64)self.outputSampleRate numberOfChannels:(UInt32)self.outputNumberOfChannels error:nil];
 
-    if (! engine.enableMicrophone) {
-        return noErr;
+    if (! [self.inputQueue begin:error] || ! [self.outputQueue begin:error]) {
+        return NO;
     }
-
-    AudioBufferList bufferList;
-    bufferList.mNumberBuffers = 1;
-    bufferList.mBuffers[0].mNumberChannels = kNumberOfInputChannels;
-    bufferList.mBuffers[0].mData = NULL;
-    bufferList.mBuffers[0].mDataByteSize = inNumberFrames * sizeof(SInt16) * kNumberOfInputChannels;
-
-    OSStatus status = _AudioUnitRender(engine.ioUnit,
-                                       ioActionFlags,
-                                       inTimeStamp,
-                                       inBusNumber,
-                                       inNumberFrames,
-                                       &bufferList);
-
-    TPCircularBufferProduceBytes(&engine->_inputBuffer,
-                                 bufferList.mBuffers[0].mData,
-                                 bufferList.mBuffers[0].mDataByteSize);
-
-    int32_t availableBytesToConsume;
-    void *tail = TPCircularBufferTail(&engine->_inputBuffer, &availableBytesToConsume);
-    int32_t minimalBytesToConsume = kSampleCount * kNumberOfInputChannels * sizeof(SInt16);
-
-    int32_t cyclesToConsume = availableBytesToConsume / minimalBytesToConsume;
-
-    for (int32_t i = 0; i < cyclesToConsume; i++) {
-        NSError *error;
-        [engine.toxav sendAudioFrame:tail
-                         sampleCount:kSampleCount
-                            channels:kNumberOfInputChannels
-                          sampleRate:engine.inputSampleRate
-                            toFriend:engine.friendNumber
-                               error:&error];
-        TPCircularBufferConsume(&engine->_inputBuffer, minimalBytesToConsume);
-        tail = TPCircularBufferTail(&engine->_inputBuffer, &availableBytesToConsume);
+    else {
+        return YES;
     }
```

Comment:
No need for else here, you can simply `return NO;`. (it is a matter of style, but I prefer early returns without "else" part).